File: algorithms/rs/20.有效的括号.rs
/*
 * @lc app=leetcode.cn id=20 lang=rust
 *
 * [20] 有效的括号
 */

// @lc code=start
impl Solution {
    pub fn is_valid(s: String) -> bool {
        let chars: Vec<char> = s.chars().collect();
        let mut stack = vec![];
        for i in 0..chars.len() {
            match chars[i] {
                '(' => stack.push(')'),
                '[' => stack.push(']'),
                '{' => stack.push('}'),
                ch => {
                    if (stack.is_empty() || stack.pop().unwrap() != ch) {
                        return false;
                    }
                }
            }
        }
        stack.is_empty()
    }
}
// @lc code=end

File: algorithms/golang/20.有效的括号.go
/*
 * @lc app=leetcode.cn id=20 lang=golang
 *
 * [20] 有效的括号
 */

// @lc code=start
func isValid(s string) bool {
	hash := map[byte]byte{
		'{': '}',
		'[': ']',
		'(': ')',
	}
	stack := []byte{}
	for _, ch := range []byte(s) {
		val, ok := hash[ch]
		if ok {
			stack = append(stack, val)
		} else if len(stack) == 0 || stack[len(stack)-1] != ch {
			return false
		} else {
			stack = stack[0 : len(stack)-1]
		}
	}
	return len(stack) == 0
}

// @lc code=end
